File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    queue <int> q;
    int answer = 0;
    int bridgeOn=0;
    int i=0;
    
    while(1){
        answer++;
        
        if(q.size()==bridge_length){
            bridgeOn-=q.front();
            q.pop();
        }
        if(bridgeOn+truck_weights[i]<=weight){
            
            if(i==truck_weights.size()-1){
                answer+=bridge_length;
                break;
            }
            q.push(truck_weights[i]);
            bridgeOn+=truck_weights[i];
            i++;
        }
        else{
            q.push(0);
        }
    }
    return answer;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
    
        Queue<Integer> q = new LinkedList<>();
        int idx=0, time=0, sum=0;
        
        while(true){
            if(idx == truck_weights.length){
                break;
            }
            
            if(q.isEmpty()){
                q.add(truck_weights[idx]);
                sum += truck_weights[idx];
                idx++;
                time++;
                continue;
                
            } else if(q.size() < bridge_length){
                if(sum + truck_weights[idx] <= weight){
                    q.add(truck_weights[idx]);
                    sum += truck_weights[idx];
                    idx++;
                    time++;
                    
                } else {
                    q.add(0);
                    time++;
                }
                continue;
                
            } 
            
            if(q.size() == bridge_length){
                sum -= q.poll();
            }
        }
    
        return time+bridge_length;
    }
}